ext/date/php_date.c

```diff
@@ -4108,30 +4108,49 @@ static int date_period_initialize(timelib_time **st, timelib_time **et, timelib_
 /* {{{ Creates new DatePeriod object. */
 PHP_METHOD(DatePeriod, __construct)
 {
-	php_period_obj   *dpobj;
-	php_date_obj     *dateobj;
-	zval *start, *end = NULL, *interval;
-	zend_long  recurrences = 0, options = 0;
-	char *isostr = NULL;
-	size_t   isostr_len = 0;
+	php_period_obj *dpobj;
+	php_date_obj *dateobj;
+	zend_object *start_obj;
+	zend_string *start_str;
+	zend_object *interval_obj = NULL;
+	zend_bool interval_is_null = 1;
+	zend_object *end_obj = NULL;
+	zend_long recurrences = 0;
+	zend_bool end_is_null = 1;
+	zend_long options = 0;
+	char *isostr;
+	size_t isostr_len;
 	timelib_time *clone;
 	zend_error_handling error_handling;
 
 	zend_replace_error_handling(EH_THROW, NULL, &error_handling);
-	if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS(), "OOl|l", &start, date_ce_interface, &interval, date_ce_interval, &recurrences, &options) == FAILURE) {
-		if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS(), "OOO|l", &start, date_ce_interface, &interval, date_ce_interval, &end, date_ce_interface, &options) == FAILURE) {
-			if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS(), "s|l", &isostr, &isostr_len, &options) == FAILURE) {
-				php_error_docref(NULL, E_WARNING, "This constructor accepts either (DateTimeInterface, DateInterval, int) OR (DateTimeInterface, DateInterval, DateTime) OR (string) as arguments.");
-				zend_restore_error_handling(&error_handling);
-				return;
-			}
-		}
-	}
+	ZEND_PARSE_PARAMETERS_START(1, 4)
+		Z_PARAM_STR_OR_OBJ_OF_CLASS(start_str, start_obj, date_ce_interface)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_OBJ_OF_CLASS_OR_LONG_OR_NULL(interval_obj, date_ce_interval, options, interval_is_null)
+		Z_PARAM_OBJ_OF_CLASS_OR_LONG_OR_NULL(end_obj, date_ce_interface, recurrences, end_is_null)
+		Z_PARAM_LONG(options)
+	ZEND_PARSE_PARAMETERS_END();
```

Comment:
We should restore error handling if ZPP fails. I think somewhere I saw an example for that, but I couldn't find it now.

Reply:
You can move the zend_replace_error_handling call below here, it should not be active during zpp.